articles/budougumi0617.re

```diff
@@ -1,15 +1,103 @@
+= GoにおけるSOLID
 
+@<tt>{@budougumi0617}です。
+#@# textlint-disable
+オブジェクト指向設計の原則として@<kw>{SOLIDの原則}@<fn>{solid}（@<tti>{the SOLID principles}）があることはみなさんご存じかと思います。
+#@# textlint-enable
+本章ではSOLIDの原則にのっとったGoの実装について考えます。
 
-= textlint 検証用文書
+//footnote[solid][@<href>{https://ja.wikipedia.org/wiki/SOLID}]
 
-はじめに。
 
-初めてのGo言語。
-それならばGo言語。
+== @<kw>{SOLIDの原則}（@<tti>{the SOLID principles}）
+SOLIDの原則は次の用語リストに挙げた5つのソフトウェア設計の原則の頭文字をまとめたものです。
+これらの原則は、ソフトウェアをより理解しやすく、より柔軟でメンテナンス性の高いものにするために考案されました。
+書籍としては、「アジャイルソフトウェア開発の奥義」@<fn>{amzn_agile}でまとめられたのが最初です。
 
-始めてみようGo言語。
+//footnote[amzn_agile][@<href>{https://www.amazon.co.jp/dp/4797347783}]
 
-@<code>{git commit --amend}はエラーにしてほしくない。（デフォルトで無視される）
-@<tt>{git commit --amend}はエラーにしてほしくない。
 
-ほかの言語。
+#@# textlint-disable
+ : @<kw>{単一責任の原則}（@<kw>{SRP}, @<tti>{Single responsibility principle}）
+    クラスを変更する理由は1つ以上存在してはならない。@<fn>{agile_srp}
+ : @<kw>{開放閉鎖の原則}（@<kw>{OCP}, @<tti>{Open–closed principle}）
+    ソフトウェアの構成要素構成要素（クラス、モジュール、関数など）は拡張に対して開いて（オープン: Oepn）いて、修正に対して閉じて（クローズド: Closed）いなければならない。@<fn>{agile_ocp}
+ : @<kw>{リスコフの置換原則}（@<kw>{LSP}, @<tti>{Liskov substitution principle}）
+    @<tt>{S}型のオブジェクト@<tt>{o1}の各々に、対応する@<tt>{T}型のオブジェクト@<tt>{o2}が1つ存在し、@<tt>{T}を使って定義されたプログラム@<tt>{P}に対して@<tt>{o2}の代わりに@<tt>{o1}を使っても@<tt>{P}の振る舞いが変わらない場合、@<tt>{S}は@<tt>{T}の派生型であると言える。@<fn>{agile_lsp}
+ : インターフェイス分離の原則（@<kw>{ISP}, @<tti>{Interface segregation principle}）
+    クライアントに、クライアントが利用しないメソッドへの依存を強制してはならない。@<fn>{agile_isp}
+ : @<kw>{依存性逆転の原則}（@<kw>{DIP},  @<tti>{Dependency inversion principle}）
+    上位のモジュールは下位のモジュールに依存してはならない。どちらのモジュールも「抽象」に依存すべきである。「抽象」は実装の詳細に依存してはならない。実装の詳細が「抽象」に依存すべきである。@<fn>{agile_dip}
+#@# textlint-enable
+
+
+
+//footnote[agile_srp][アジャイルソフトウェア開発の奥義 第2版 8.1より]
+//footnote[agile_ocp][アジャイルソフトウェア開発の奥義 第2版 9.1より]
+//footnote[agile_lsp][アジャイルソフトウェア開発の奥義 第2版 10.1より]
+//footnote[agile_isp][アジャイルソフトウェア開発の奥義 第2版 12.3より]
+//footnote[agile_dip][アジャイルソフトウェア開発の奥義 第2版 11.1より]
+
+== Goとオブジェクト指向プログラミング
+本題に入る前に、Goとオブジェクト指向プログラミングの関係について考えてみます。
+#@# textlint-disable
+Go公式サイトには@<kw>{Frequently Asked Questions (FAQ)}@<fn>{q_and_a}という「よくある質問の答え」ページがあります。
+#@# textlint-enable
+この中に@<i>{Is Go an object-oriented language?}（Goはオブジェクト指向言語ですか？）という質問に対する公式見解が述べられています。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Disallow to use "？". (ja-technical-writing/no-exclamation-question-mark)